### PR TITLE
Feature erddap query

### PR DESCRIPTION
**Pull request by kmilo9999**
connect to erddap and retrieve data to create volume viewer dataset

### CODE REVIEW

**Comment by kmilo9999** [comment]
@gtdang @mirestrepo  just a quick review here.

File: volume_data_converter/erddap_query.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        ## check the file name format and extension
+        folder_path, filename = os.path.split(output_file_path)
+        if not os.path.isdir(folder_path):
+            raise (
+                f" Could not find output path: {output_file_path}. Check first argument"
+            )
+        name, ext = os.path.splitext(filename)
+        if ext != ".nc":
+            ext = ".nc"
+        output_file_path = os.path.join(folder_path, name + ext)
```
Comment (gtdang) [inline volume_data_converter/erddap_query.py:62]: This if/else clause is pretty hard to follow. Some additional comments might be useful. If I understand it correctly....

- The `if` handles if the input is only a path to a directory and will create a default file.
- The `else` handles if the input is a path + filename.
- It is worth noting that the `else` only handles a filename with extension. If no extension the code will crash? 

Perhaps it would be more clear to have it structured as: 

1. `if` an existing directory 
2. `elif` a file with extension - You can check this by `'.' in output_file_path` 
3. `else` raise input error

File: volume_data_converter/erddap_query.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    # set up connection properties
+    erddap_obj = ERDDAP(
+        server=erddap_connection_object["server"],
+        protocol=erddap_connection_object["protocol"],
+    )
+    erddap_obj.dataset_id = erddap_connection_object["dataset_id"]
+
+    ## initialize and retrive erddap information
+    erddap_obj.griddap_initialize()
+    typer.echo(f"Connect to server: {erddap_connection_object['server']}")
```
Comment (gtdang) [inline volume_data_converter/erddap_query.py:73]:
```suggestion
    typer.echo(f"Connecting to server: {erddap_connection_object['server']}")
```

File: volume_data_converter/erddap_query.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    typer.echo(f"Looking up dataset id : {erddap_connection_object['dataset_id']}")
+
+    ## set time and variables constraints
+    erddap_constrainst_object = configuration["erddap_constrainst"]
```
Comment (gtdang) [inline volume_data_converter/erddap_query.py:77]: Should these variables should be spelled "constraints"?

**Review by gtdang — COMMENTED** [review]
Let me know you have any questions!

File: volume_data_converter/erddap_query.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    """
+
+    print(f"configuration_file_path {erddap_configuration_file}")
+    try:
+        conf_file = open(erddap_configuration_file, "r")
+    except OSError:
+        raise (f" Could not open/read file: {erddap_configuration_file} ")
+
+    with conf_file:
```
Comment (mirestrepo) [inline volume_data_converter/erddap_query.py:39]: The purpose of this syntax is that the file gets closed after loading, but your `try` statement above is opening the file and it's not closing it, so you'll have a dangling file handler. I think the try/except is unnecessary. If the file doesn't open, you'lll get an error anyways

File: volume_data_converter/erddap_query.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    typer.echo(f"Connect to server: {erddap_connection_object['server']}")
+    typer.echo(f"Looking up dataset id : {erddap_connection_object['dataset_id']}")
+
+    ## set time and variables constraints
+    erddap_constrainst_object = configuration["erddap_constrainst"]
+
+    check_variables = all(
+        item in erddap_obj.variables for item in erddap_constrainst_object["variables"]
+    )
+    if not check_variables:
```
Comment (mirestrepo) [inline volume_data_converter/erddap_query.py:82]: what is this doing? just asking to verify visually?

File: volume_data_converter/erddap_query.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    ds.to_netcdf(output_file_path)
+    spinner.stop()
+    typer.echo(f"Data downloaded to file {output_file_path}")
+
+
+def main():
+    app()
+
+
+# if __name__ == "__main__":
```
Comment (mirestrepo) [inline volume_data_converter/erddap_query.py:108]: Remove

File: volume_data_converter/erddap_query.py (hunk 1 above)
Comment (mirestrepo) [reply volume_data_converter/erddap_query.py:62]: I left a comment also related to this. I would vote for not doing so many checks and letting the errors just surfacing naturally

File: volume_data_converter/create_volume_viewer_data.py [hunk 7]
```diff
@@ -6,14 +6,27 @@
...
+constants_file_path = os.path.join(
+    Path(__file__).absolute().parent, "config", "constants.json"
+)
+
 
 def fprintf(stream, format_spec, *args):
     stream.write(format_spec % args)
 
 
+def getVariableValOrDefault(configuration_file, nc_file_handler, variable_name: str):
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: I deleted the function. I dodnt realize there is already a function in python that query's data from a dictionary  and sets up a default value in case it doesnt find it.

File: volume_data_converter/create_volume_viewer_data.py [hunk 8]
```diff
@@ -26,6 +39,10 @@ def createOsomData(
         [],
         help=" List of time frames to convert to raw. By default is None: it will convert all the time frames in a .nc file",
     ),
+    layer: str = typer.Option(
+        "all",
+        help=" Layers of the osom model this nc files maps to. Options: all, surface, bottom",
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: done

File: volume_data_converter/create_volume_viewer_data.py [hunk 9]
```diff
@@ -36,10 +53,24 @@ def createOsomData(
...
     """
+
+    try:
+        conf_file = open(constants_file_path, "r")
+    except OSError:
+        typer.echo(" Could not open/read file: " + constants_file_path)
+        sys.exit()
+
+    with conf_file:
+        configuration_file = json.load(conf_file)
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: I remove this try/catch block. Let's follow the philosophy that if the system doesnt find the file, it throws an exception either way explaining what is happening.

File: volume_data_converter/create_volume_viewer_data.py [hunk 10]
```diff
@@ -52,15 +83,55 @@ def createOsomData(
...
-    ocean_time = nc_dataFile.variables["ocean_time"][:]
-    ocean_time_properties = nc_dataFile.variables["ocean_time"]
+    data_properties = nc_dataFile.variables[data_descriptor]
+    data_dimensions = data_properties.dimensions
+
+    ## zeta needs a different treatment form the other variables. If it's not in the configuration file, then build it from the descriptor's data itself
+    if "zeta" in nc_dataFile.variables:
+        zeta = nc_dataFile.variables["zeta"][:]
+    else:
+        zeta = np.zeros(shape=(data.shape))
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: done

File: volume_data_converter/create_volume_viewer_data.py [hunk 11]
```diff
@@ -36,10 +53,24 @@ def createOsomData(
...
+
+    with conf_file:
+        configuration_file = json.load(conf_file)
+
     # Default factors/scalers
     verticalLevels = 15
     downscaleFactor = 2
 
+    # check layers
+    if layer not in ["all", "surface", "bottom"]:
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py:59]: I applied `lower()`.

File: volume_data_converter/create_volume_viewer_data.py [hunk 12]
```diff
@@ -52,15 +83,55 @@ def createOsomData(
...
+        zeta = nc_dataFile.variables["zeta"][:]
+    else:
+        zeta = np.zeros(shape=(data.shape))
+
+    # check if it's a volume or a single slice. Make the corrsponding transformation to 3D array
+
+    if layer == "all" and "s_rho" not in data_dimensions:
+        raise Exception("current dataset does not support elevation layers")
+
+    elif layer != "all" and "s_rho" not in data_dimensions:
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: I rewrote this logic and left comments. let me know if it's clear enough.

File: volume_data_converter/create_volume_viewer_data.py [hunk 13]
```diff
@@ -114,9 +185,11 @@ def createOsomData(
                         t_data = data[time_t, :, x, y]
                         t_data = np.ma.squeeze(t_data)
                         idx = ~t_data.mask
+
                         if np.sum(idx.astype(int)) > 0:
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py:177]: `idx = ~t_data.mask` returns a vector of booleans. `np.sum(idx.astype(int))` checks if the whole array has at least 1 true value. It works so far for both multi and single layer datasets. Maybe there is a better way to do it. I'm open to suggestions.

File: volume_data_converter/erddap_query.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    """
+
+    print(f"configuration_file_path {erddap_configuration_file}")
+    try:
+        conf_file = open(erddap_configuration_file, "r")
+    except OSError:
+        raise (f" Could not open/read file: {erddap_configuration_file} ")
+
+    with conf_file:
+        configuration = json.load(conf_file)
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py]: I remove this try/catch block. Let's follow the philosophy that if the system doesnt find the file, it throws an exception either way explaining what is happening.

File: volume_data_converter/erddap_query.py (hunk 1 above)
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py:56]: 1. In the first `if` I am checking if the directory exists. Just adding a message indicating that a files will be created in that folder and its name. I think the prospect users of this app need feedback from the app telling them what is happening.
2.  What is wrong by using the built-in function `os.path.splitext` ?  If there no extension, the variable `ext` is empty and it will enter into the `if ext != ".nc"`  that adds the extension. It wont crash.
3. In the `elif ` ( file with extension case ), I still need to check if in the given path, the parent folder exists.

File: volume_data_converter/erddap_query.py (hunk 2 above)
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py]: done

File: volume_data_converter/erddap_query.py (hunk 3 above)
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py]: done

File: requirements.txt [hunk 15]
```diff
@@ -1,29 +1,63 @@
...
-click==8.0.3
-colorama==0.4.4
-cycler==0.11.0
-fonttools==4.29.1
-imagecodecs==2021.11.20
-imagesize==1.3.0
-importlib-metadata==4.10.1
-kiwisolver==1.3.2
-matplotlib==3.5.1
+appdirs==1.4.4
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply requirements.txt]: I can remove the `requeriments.txt` file. It is a residual from when the project didnt use poetry.

File: volume_data_converter/create_volume_viewer_data.py [hunk 16]
```diff
@@ -6,14 +6,27 @@
 from volume_data_converter import scoord_du_new2 as scoor_du
 from typing import List
 
+import json
+import sys
+from pathlib import Path
 
 app = typer.Typer()
 
+constants_file_path = os.path.join(
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: I removed it and placed as the default value of the `erddap_query` function.

File: volume_data_converter/create_volume_viewer_data.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: I took it out.

File: volume_data_converter/create_volume_viewer_data.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: We talked about this. The non multi layer nc files have no knowledge of their position in space or where they should be map. That info is only available through the variable name in erddap. It is possible to create the two volumes and merge them into a single volume visualization. 
Let me do the layer merge in another PR. This one has some issues that I fixed in another ready to go PR waiting for this one to get approve.

File: volume_data_converter/erddap_query.py (hunk 4 above)
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py]: I removed the exception.

File: volume_data_converter/erddap_query.py (hunk 5 above)
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py:77]: Checking the requested variables in the local `erddap_configurarion.json`  file exists in the erddap database.

File: volume_data_converter/erddap_query.py (hunk 6 above)
Comment (kmilo9999) [reply volume_data_converter/erddap_query.py]: Done.

File: volume_data_converter/create_volume_viewer_data.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (gtdang) [reply volume_data_converter/create_volume_viewer_data.py:177]: Ah ok. You could do `if any([boolean array])` then.

[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: done

File: volume_data_converter/erddap_query.py (hunk 1 above)
Comment (gtdang) [reply volume_data_converter/erddap_query.py:56]: Ah I see, you're right.

**Review by gtdang — COMMENTED** [review]
Sorry for taking to so long! I think it's looking pretty good. Just have some minor edits and questions.

File: volume_data_converter/create_volume_viewer_data.py [hunk 17]
```diff
@@ -52,15 +71,55 @@ def createOsomData(
...
+    theta_s = nc_dataFile.variables.get("theta_s",osom_configuration_dicc["theta_s"])
+    theta_b = nc_dataFile.variables.get("theta_b",osom_configuration_dicc["theta_b"])
+    hc = nc_dataFile.variables.get("hc",osom_configuration_dicc["hc"])
+
+    time_variable_name = osom_configuration_dicc["time_variable"]
+    if time_variable_name not in nc_dataFile.variables:
+        raise Exception("Time variable not found")
+
+    ocean_time = nc_dataFile.variables[time_variable_name][:]
+    ocean_time_properties = nc_dataFile.variables[time_variable_name]
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: I changed the lines to reflect on what they do.
One is the header with the properties of that variable. The other one is the list of each ocean time for each time-frame.

File: volume_data_converter/create_volume_viewer_data.py [hunk 18]
```diff
@@ -52,15 +71,55 @@ def createOsomData(
...
+    else :
+        # no elevation data. Map the data to a empty block of 'verticalayers' layers.
+        # Our current cases are surface and bottom. They map to layer 0 adn 14 respectively 
+        new_data = np.ma.zeros(
+            (data.shape[0], verticalLevels, data.shape[1], data.shape[2]),
+            dtype=data.dtype,
+        )
+        data_slice = 0  # bottom by default
+        if layer == "surface":
+            data_slice = verticalLevels - 1
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: Donde

File: volume_data_converter/create_volume_viewer_data.py [hunk 19]
```diff
@@ -52,15 +71,55 @@ def createOsomData(
...
+    # check if it's a volume or a single slice. Make the corrsponding transformation to 3D array
+
+    if layer.lower() == "all":
+        # s_rho dimension implies that the data is distributed on multiple layers
+        # all is the default value. Check if the dataset is multilayer or not.
+        if "s_rho" not in data_dimensions:
+            raise Exception("current dataset does not support elevation layers")
+    else :
+        # no elevation data. Map the data to a empty block of 'verticalayers' layers.
+        # Our current cases are surface and bottom. They map to layer 0 adn 14 respectively
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply volume_data_converter/create_volume_viewer_data.py]: Done

**Review by gtdang — APPROVED** [review]
Looks good to me!

File: volume_data_converter/create_volume_viewer_data.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (mirestrepo) [reply volume_data_converter/create_volume_viewer_data.py]: I still see this global in the current changes. Just define it in `createOsomData`

**Review by mirestrepo — APPROVED** [review]
I left one minor last question. For another day, but convention for function names in Python is snake case https://peps.python.org/pep-0008/#function-and-variable-names
